### ISSUE TITLE
Configure NFS version for Primary Storage

### ISSUE THREAD

**Issue by wido** [apache/cloudstack#4482]
#### ISSUE TYPE
 * Improvement Request

##### COMPONENT NAME
~~~
NFS Primary Storage
~~~

##### CLOUDSTACK VERSION
~~~
ALL
~~~

##### CONFIGURATION
- Primary Storage
- NFS

##### OS / ENVIRONMENT
- Ubuntu / CentOS

##### SUMMARY
NFS Primary Storage mounts are handled by libvirt.

Currently libvirt defaults to NFS version 3 when mounting while it does support NFS version 4 if provided in the XML definition: https://libvirt.org/formatstorage.html#StoragePoolSource

~~~
  <source>
    <host name='localhost'/>
    <dir path='/var/lib/libvirt/images'/>
    <format type='nfs'/>
    <protocol ver='4'/>
  </source>
~~~

Maybe pass the argument 'nfsvers' to the URL provided to the Management Server and then pass this down to the Hypervisors which generate the XML for libvirt.

##### EXPECTED RESULTS
~~~
Primary Storage is mounted over NFS version 4
~~~

**Comment by DaanHoogland** [comment]
looks great, @wido . plan to put it behind a config option? or in the create primary call? (/me having the impertinence of marking it for 4.16)

**Comment by wido** [comment]
> looks great, @wido . plan to put it behind a config option? or in the create primary call? (/me having the impertinence of marking it for 4.16)

See: https://cloudstack.apache.org/api/apidocs-4.14/apis/createStoragePool.html

We provide the URL in this form:

<pre>nfs://my-nfs-server/exports/ps01</pre>

We could change that URL by setting:

<pre>nfs://my-nfs-server/exports/ps01?nfsvers=4</pre>

**Comment by DaanHoogland** [comment]
does that mean that it can be upgraded to version 4 by updating the URL, appending the nfsvers argument? (url is no argument for updateStoragePool now: https://cloudstack.apache.org/api/apidocs-4.14/apis/updateStoragePool.html)
or is any other mech envisioned?

**Comment by wido** [comment]
> does that mean that it can be upgraded to version 4 by updating the URL, appending the nfsvers argument? (url is no argument for updateStoragePool now: https://cloudstack.apache.org/api/apidocs-4.14/apis/updateStoragePool.html)
> or is any other mech envisioned?

I don't think we can do and want to do this for existing primary storages. People could hack the DB and modify it if they really want to.

Otherwise we only support it when adding a new Primary Storage. We just need to parse the URL and store the nfs version in the DB and pass it all the way down to the XML we provide to libvirt.

**Comment by DaanHoogland** [comment]
copy

**Comment by nvazquez** [comment]
@wido @DaanHoogland there is a global setting to control the NFS version for secondary storage 'secstorage.nfs.version' perhaps it's easier to include a new global setting for primary storage and pass the version on the storage pool parameters down to the XML? cc @weizhouapache

**Comment by GabrielBrascher** [comment]
@nvazquez this could be an option indeed

**Comment by rhtyd** [comment]
@GabrielBrascher @wido @nvazquez @sureshanaparti @davidjumani @shwstppr @Pearl1594 @DaanHoogland @weizhouapache - any update on this, who's assigned/working on this? Or should we change milestone? Thanks.

**Comment by nvazquez** [comment]
Moved to the 4.17 milestone

**Comment by nvazquez** [comment]
Hi @wido @GabrielBrascher is there any update on this?

**Comment by GabrielBrascher** [comment]
@nvazquez not yet, this is still on our "backlog".

**Comment by theexplorist** [comment]
> Ok - moving it tentatively to the 4.18 milestone

Hi @nvazquez I wasn't well the last 2 weeks of past month sustaining a back injury, it will helpful if we can have a small connect regarding the project. I have mailed you if we can connect the coming week then kindly let me know.
Thanks

**Comment by shwstppr** [comment]
We do not have any work in progress for this at the moment so moving it out of the 4.19.0.0 milestone

**Comment by rohityadavcloud** [comment]
This would be fixed by https://github.com/apache/cloudstack/pull/8947

**Comment by sureshanaparti** [comment]
Fixed in #8947
